### PR TITLE
Cart models

### PR DESCRIPTION
**Pull request by braimor**
Cart and LineItem models.

### CODE REVIEW

File: app/models/cart.rb [hunk 1]
```diff
@@ -0,0 +1,2 @@
+class Cart < ApplicationRecord
+end
```
Comment (diebas) [inline app/models/cart.rb:13]: before_save :update_total

def update_total
   ...
end

File: app/models/line_item.rb [hunk 2]
```diff
@@ -0,0 +1,4 @@
+class LineItem < ApplicationRecord
+  belongs_to :product
```
Comment (diebas) [inline app/models/line_item.rb:2]: add total value, same for cart. with a default value of 0 and not nullable

File: app/views/carts/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,25 @@
+<p id="notice"><%= notice %></p>
```
Comment (diebas) [inline app/views/carts/index.html.erb]: delete views

File: app/controllers/carts_controller.rb [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  # GET /carts or /carts.json
+  def index
+    @carts = Cart.all
+  end
+
+  # GET /carts/1 or /carts/1.json
+  def show
+  end
+
+  # GET /carts/new
```
Comment (diebas) [inline app/controllers/carts_controller.rb]: delete unused controller actions

File: app/controllers/carts_controller.rb [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+  # GET /carts/1/edit
+  def edit
+  end
+
+  # POST /carts or /carts.json
+  def create
+    @cart = Cart.new(cart_params)
+
+    respond_to do |format|
+      if @cart.save
```
Comment (diebas) [inline app/controllers/carts_controller.rb]: use same format as the rest of the app, use other controllers as example

File: app/models/cart.rb [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+    def update_total
+        total = 0
+        line_items.each do |line_item|
+            total += line_item.total_value
+        end
+        total
+    end
+
+    def total_value
```
Comment (diebas) [inline app/models/cart.rb]: is this method still needed?

File: app/models/cart.rb [hunk 7]
```diff
@@ -0,0 +1,16 @@
+class Cart < ApplicationRecord
+    has_many :line_items, dependent: :destroy
+    before_save :update_total
+
```
Comment (diebas) [inline app/models/cart.rb:5]: `validates :total, presence: true 
same for line_items

File: app/models/line_item.rb [hunk 8]
```diff
@@ -0,0 +1,8 @@
...
+  belongs_to :cart
+
+  def total_value
```
Comment (diebas) [inline app/models/line_item.rb:7]: add a callback as we did with the carts model to update the total value in the database

File: app/models/product.rb [hunk 9]
```diff
@@ -4,4 +4,6 @@ class Product < ApplicationRecord
   has_one_attached :image
   include Pagy::Frontend
   validates_presence_of :price, :name
+  has_many :line_items
```
Comment (diebas) [inline app/models/product.rb]: `has_many :line_items, dependent: :destroy` 
This is to make sure users don't continue having the product in their cart if that product was deleted

File: app/models/product.rb [hunk 10]
```diff
@@ -4,4 +4,6 @@ class Product < ApplicationRecord
...
+  before_destroy :ensure_not_referenced_by_any_line_item
```
Comment (diebas) [inline app/models/product.rb]: what does this method do?

File: config/routes.rb [hunk 11]
```diff
@@ -1,6 +1,8 @@
 # frozen_string_literal: true
 
 Rails.application.routes.draw do
+  resources :line_items
```
Comment (diebas) [inline config/routes.rb]: resources adds all methods for the controller. is a good practice to just add only the routes for the actions you created.
`resources :line_items, only: [:new, :create]`

same for carts

File: app/models/product.rb (hunk 10 above)
Comment (braimor) [reply app/models/product.rb]: it's a not defined method that will ensure that a product in a cart can't be deleted.

File: app/controllers/carts_controller.rb [hunk 12]
```diff
@@ -0,0 +1,26 @@
+class CartsController < ApplicationController
+  before_action :set_cart, only: %i[show edit update destroy]
+
+  def new
+    @cart = Cart.new
+  end
+
+  def create
+    @cart = Cart.new(cart_params)
```
Comment (diebas) [inline app/controllers/carts_controller.rb]: you are not persisting the record on the database, use create! intead of new

File: app/controllers/carts_controller.rb [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+  end
+
+  def update
+    current_cart.update(cart_params)
```
Comment (diebas) [inline app/controllers/carts_controller.rb]: use update! , the bang is used to raise exceptions when the action fails

File: app/controllers/line_items_controller.rb [hunk 14]
```diff
@@ -0,0 +1,21 @@
+class LineItemsController < ApplicationController
+  before_action :set_line_item, only: %i[show edit update destroy]
```
Comment (diebas) [inline app/controllers/line_items_controller.rb]: be carefull, you are including actions you didn't defined yet.

File: app/controllers/line_items_controller.rb [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+
+  def new
+    @line_item = LineItem.new
+  end
+
+  def create
+    @line_item = LineItem.new(line_item_params)
```
Comment (diebas) [inline app/controllers/line_items_controller.rb]: same as with cart

**Review by diebas — APPROVED** [review]

**Pull request merged**